### ISSUE TITLE
Add QUIC support for Minetest

### ISSUE THREAD

**Issue by paradust7** [luanti-org/luanti#15261]
### Problem

In the current Minetest protocol, servers are unauthenticated and the connection is unencrypted. It is therefore always possible for a MITM to hijack or snoop on a connection after a successful login.

Additionally, the protocol uses UDP, and so must implement features such as reliability (re-transmission), flow control, packet re-ordering, and channels. The implementations of these features are necessarily simple and unlikely to be optimal, compared to implementations by well-funded network engineering teams.

### Solutions

Minetest can make use of QUIC as an alternative to UDP for establishing connections to a server.

### How QUIC works

QUIC uses the same security mechanism as https, so a connection to a server with a host name will be secure by default. The server operator will need to generate a certificate for their domain or subdomain, e.g., using Let's Encrypt or other provider.

Connecting to a server by IP will still be encrypted, but using an unverified certificate (which could be self-signed). The user will be warned that this is insecure. The server's certificate will be cached, so that subsequent connects expect the same certificate. If the certificate changes, the user will be loudly warned. This can't prevent a MITM attack from occurring on the very first connect, but it can detect and prevent future attacks.

After the TLS handshake, QUIC connections remain encrypted using a negotiated ephemeral cipher. The details
are handled by the QUIC protocol.

### How Minetest can use QUIC

QUIC provides channels (streams), reliable transmission, and in-order packet delivery. Minetest's protocol currently emulates these things over UDP. It will be possible to use QUIC for all of these features, in addition to raw datagrams (rfc 9221) for plain UDP-like delivery.

It *may* also be possible to have Minetest serve both the UDP protocol and QUIC protocol on the same port. (Investigation is needed)

**Comment by paradust7** [comment]
As usual, I am volunteering myself to implement this, and will be upstreaming if possible. This issue is to track comments and progress. :tada:

**Comment by sfan5** [comment]
related: #10206, #14970
tangentially related to #4184 (this would also rid us of our terrible netcode)

> The server operator will need to generate a certificate for their domain or subdomain, e.g., using Let's Encrypt or other provider.
> Connecting to a server by IP will still be encrypted, but using an unverified certificate (which could be self-signed). The user will be warned that this is insecure. [...]. If the certificate changes, the user will be loudly warned.

In related discussions the consensus more or less was that requiring certificate PKI is bad UX for the server owner, player or even both.

> It may also be possible to have Minetest serve both the UDP protocol and QUIC protocol on the same port. (Investigation is needed)

IMO this is quite important to have for a streamlined upgrade experience (and also to prevent admins from making mistakes where only one of the protocols works).
This is in principle always possible as long as the QUIC implementation supports abstracting socket access and doesn't insist directly touching sockets on its own. Because our own protocol will have a different magic number.

**Comment by paradust7** [comment]
> In related discussions the consensus more or less was that requiring certificate PKI is bad UX for the server owner, player or even both.

For dedicated server owners, I don't think it is too much to ask. If they already have a website at the same domain, they would only need to copy the existing certificate. Otherwise, an invocation of "certbot" would be able to generate one. I will be sure to write up detailed instructions.

Luckily, the player doesn't need to do anything with certificates. Everyone else running a server can use a default self-signed certificate.

Later, it would be possible to add support for secure reverse tunneling proxies. This would enable non-tech-savvy players to host games behind a NAT and without a certificate, while maintaining full security. The UX for this would be:
- Player starts a game, selects Host Server, selects Use Proxy "some.untrusted.host".
- Game starts, minetest produces an id + shared secret for the host to share with friends.
- Friends paste the id + shared secret into minetest and connect via the proxy.

The proxy doesn't need to be trusted, because the shared secret will be unknown to the proxy. If it tries to MITM, it will fail validation.

> IMO this is quite important to have for a streamlined upgrade experience (and also to prevent admins from making mistakes where only one of the protocols works).

Agreed

> This is in principle always possible as long as the QUIC implementation supports abstracting socket access and doesn't insist directly touching sockets on its own. Because our own protocol will have a different magic number.

Agreed.

**Comment by swagtoy** [comment]
This is a good excuse for me to finally learn QUIC, but let me QUICly drop some late-night thoughts that comes to mind.

Why do we NEED to encrypt all of the games packets? For embedded devices, that could make some extra useless work, considering how fast QUIC encryption is. I can understand encrypted chat logs, maybe some modchannel stuff, given the modder so pleases, but for every packet?

I mean, man, id be mad if I got my base griefed because the grayhair at Mcdonalds picked up some peculiar packets. For MITM attacks I can understand, but a WPA like handshake at the beginning could do the trick to prevent this, not overall encryption. If a player so direly fears having his Mese shards stolen by the college wifi pest, shall we make it an option?

(yes, encryption isnt heavy, but id prefer compression over encryption here, some people play on dialup toasters being oven baked in a hot room)

**Comment by paradust7** [comment]
> Why do we NEED to encrypt all of the games packets? For embedded devices, that could make some extra useless work, considering how fast QUIC encryption is. I can understand encrypted chat logs, maybe some modchannel stuff, given the modder so pleases, but for every packet?

In QUIC encryption is not optional. All traffic is encrypted, and I don't think it would be wise to tamper with this. The implementation is expected to be fast enough that this is not a problem. Even if it were possible, I don't think it would make sense to cede security decisions to individual mods. The existing protocol can still be used if someone has a need.

> (yes, encryption isnt heavy, but id prefer compression over encryption here, some people play on dialup toasters being oven baked in a hot room)

There's no conflict here. The data can be compressed before it is encrypted. This is already done in the minetest protocol for bulk node data. Compression is much more CPU intensive than encryption.

**Comment by appgurueu** [comment]
Damn, paradust beat me to it :)

>  All traffic is encrypted, and I don't think it would be wise to tamper with this.

I strongly agree. Think about it: Which packets can you safely assume will be free of the need for confidentiality *and integrity* in general? It is rather arbitrary to just pick chat and nothing else.
Essentially none of the heavier packets can be assumed to not need these things. Mapblocks? Could contain private data in metadata or even node patterns. Interactions need their integrity protected, think luablocks. Objects? Could contain private data, and again interactions may need protection of integrity.

> The implementation is expected to be fast enough that this is not a problem.

Also this. Why think about this when we can afford to just not to, *and* be better off in the integrity and confidentiality departments?

(And as paradust noted, even if we're missing something here, the old protocol remains as a fallback.)

**Comment by sfan5** [comment]
I suggest we don't start a general encryption debate here, because consensus among the core team already exists (or does it?) that we should eventually do it.

**Comment by swagtoy** [comment]
> Mapblocks? Could contain private data in metadata or even node patterns. Interactions need their integrity protected, think luablocks. Objects? Could contain private data, and again interactions may need protection of integrity.

@appgurueu Thats true, so ill leave it here. If i were to get deep then id go as far to say offering control from the Lua api, but thats incredibly user (modder) unfriendly and overkill when just encrypting everything will lightly do the job.

And if its optional? Then lets ship it ;) Nothing wrong with that

**Comment by HybridDog** [comment]
My opinion about QUIC: https://github.com/minetest/minetest/issues/10206#issuecomment-1275094664

**Comment by Darin755** [comment]
I like this idea a lot. 

Just a few notes:

 - There should be Let's encrypt integration so that server operators don't have to worry to much about renewing certs
 - There should be some way to do and verify self signed certs. Once the users accepts the self signed cert they should not need to accept again unless the cert changes. The warning message should not be to scary and should not use to much technical jargon. Maybe something like "This server failed public verification. Continue?"

Also it might be possible to use certs for client authentication. They could enter a password once and then use a client cert in the future to login. If you are writing code to do certificates anyway it shouldn't be that complicated to implement. Alternatively you could also use some sort of cookie to allow the client to authenticate without the user entering a password.

**Comment by Desour** [comment]
My opinion on QUIC: https://github.com/minetest/minetest/issues/10206#issuecomment-1225711134 part (2)
Do dtls first, and later add QUIC support. The QUIC PR then doesn't have to bother with authenticity issues. And if you use gnutls and ngtcp2 they can share code paths.

For authenticity in dtls, the simplest would be to first just authenticate the connection key with SRP. See also red's PR.

**Comment by nerzhul** [comment]
by QUIC you mean HTTP/3 support then, as it's QUIC ? This permit to ensure mTLS authentication of the server, we can encrypt every packet but what is the impact on latency, especially ojn movement packets

**Comment by Darin755** [comment]
How much work has been done on implementing this?

If there isn't much code written yet it might make sense to build this with an existing library. For instance, here is a library by Microsoft that seems pretty stable.

https://github.com/microsoft/msquic

I'm not really a cpp dev but in theory it should be a matter if just writing some filler code to map the current network code to the library. My thought is that it would be good to have a experimental version of Luanti for testing. Even if this library isn't used in the final implementation it still would be good to have a solid base instead of doing everything from scratch. I would love to test a version of Luanti that leverages QUIC.

**Comment by minlemon** [comment]
I suppose I will use this discussion to talk about QUIC-specific things. I have spent the past day looking at different QUIC libraries written in C/C++ and have come to the following conclusions:

* [msquic](https://github.com/microsoft/msquic): Written in C because they required compatibility with the win32 kernel. In my opinion, [the example](https://github.com/microsoft/msquic/blob/main/src/tools/sample/sample.c) looks far too ugly, complicated, and depends heavily on macros. The library itself is well supported. Dependencies in submodules, CMake is supported but PowerShell is the most well supported way of building it. I took a stab at making a "hello world" style program with msquic and found it difficult to do.
* [mvfst](https://github.com/facebook/mvfst): Similarly complex [example](https://github.com/facebook/mvfst/tree/main/quic/samples/echo). Written in C++ but does not make use of coroutines and requires callbacks for everything. Building is a bit of a mess and requires running a custom Python script to gather the dependencies, which fetches gzip'd files from various sources around the web (these are checked by hash). The build outputs are put into a strange temp folder. Better interface than msquic in my opinion.
* [ngtcp2](https://github.com/ngtcp2/ngtcp2): Written in C. The API can be low level if you need it but it comes with implementations for most of the low-level stuff out of the box. Allows for different backend TLS libraries to be used. Read the [library overview](https://nghttp2.org/ngtcp2/programmers-guide.html).
* [lsquic](https://github.com/litespeedtech/lsquic): The examples actually appear to be readable for this library (see [this example](https://github.com/litespeedtech/lsquic/blob/master/bin/echo_client.c), which apparently doesn't compile on Windows and hasn't been fixed). Written in C, but that might be forgivable. Requires BoringSSL as a dependency (custom fork of OpenSSL by Google, Google actually doesn't really recommend OSS projects to use it). I got build errors on the lsquic library itself and gave up after a while.
* [picoquic](https://github.com/private-octopus/picoquic): Written in C, depends on the picotls library and OpenSSL. [Example](https://github.com/private-octopus/picoquic/blob/master/sample/sample_client.c) seems potentially usable. Docs are incomplete but promising. It's branded as a "minimalist" implementation, but as you can guess by now QUIC != simple.
* [openssl](https://github.com/openssl/openssl/blob/master/README-QUIC.md): I did not find this easily, but after years of lacking support, OpenSSL now has QUIC support (3.2 client, 3.5 server). OpenSSL 3.5 was released a few weeks ago and is a LTS release. This might be our best bet and I will review example code in this library next. The [blocking-loop example](https://github.com/openssl/openssl/blob/master/demos/quic/server/server.c) has limitations but I have to guess OpenSSL has a user-friendly design once I get into it. Seems to have good documentation.

There are no other C/C++ libraries I could locate that are usable. The other ones I've crossed out are:
* [quiche (Cloudflare)](https://github.com/cloudflare/quiche): Written in Rust which has a "thin" C API. Didn't try building it.
* [quiche (Google)](https://github.com/google/quiche): "it is on the QUICHE team's roadmap to include default implementation for all platform APIs and to open-source build files" - Currently unusable/unbuildable without lots of work.
* [xquic (Alibaba)](https://github.com/alibaba/xquic): Written in C, BoringSSL dependency, demo looks complicated, connection to docs site timed out. Has changeable congestion control.
* All unmaintained projects

**Comment by appgurueu** [comment]
I have read bad things about OpenSSL QUIC support vs ngtcp2 coming from the curl author (Daniel Stenberg). See https://daniel.haxx.se/blog/2025/02/16/openssl-does-a-quic-api/ or https://mastodon.social/@bagder/114349633516219609:

> We still consider the curl backend using the OpenSSL-QUIC implementation as experimental and discourage users from using such builds in production. Now primarily because it is a performance and resource hog compared to the competition. The OpenSSL-QUIC implementation using curl has been measured up to 4 times (!) slower than ngtcp2 using up to 25 times (!) the amount of memory.
> [...]
> ngtcp2 is the world’s generic QUIC stack

**Comment by minlemon** [comment]
I appreciate the read. I have not read any other reviews of these libraries, and it's good to have multiple perspectives. I think performance is a lesser concern at the moment since in absolute terms it's probably still minimal. It would have been nicer if Daniel's comparison also included the absolute differences, like 200 KB vs 5 MB memory usage (or 1 MB vs 25 MB), and as shown at the bottom of the post, some of his criticisms and doubt about OpenSSL's implementation of features like 0RTT were resolved by OpenSSL's 3.5 release.

I wish ngtcp2's example wasn't a full HTTP client. Its docs do give some good advice like reading the RFCs [9000](https://datatracker.ietf.org/doc/html/rfc9000.html) and [9001](https://datatracker.ietf.org/doc/html/rfc9001.html) which I haven't done yet, so I have yet to be able to comment on the pull-based or push-based design choice. I think which one we'll use will come down to usability. From a brief glance it appears that OpenSSL's docs on QUIC support are better, as they provide several [in-depth guides](https://docs.openssl.org/master/man7/ossl-guide-introduction/#contents-of-the-openssl-guide) with code snippets and all. On the other hand, maybe I just need to read the RFCs and then ngtcp2 will be usable.

**Comment by minlemon** [comment]
Update after over a year. I am working on an implementation for QUIC support. Some notes:
- It supports both the web PKI (meaning you have to have a domain name and get a certificate from a certificate issuer) and local self-signed certificates (meaning you need to share a fingerprint of the certificate over a secure channel out of band, in other words, wherever you advertise your server, but you don't need to buy a domain name and don't need to look up anything from external servers).
- If the server uses a self-signed certificate, you have to add the fingerprint (essentially base64 of the SHA-256 hash of the certificate) to the "address" parameter in the UI when connecting. Maybe we can add a separate field for this, but for now it's easy to copy and paste into the Address field. Actually in general I think it would be better to have a unified field for host, port, and now fingerprint hash too... But that's my personal opinion and a UI change for a different day.
  - By the way this will look like `192.168.1.68:30000+Du1JozorVZhmsSB/eedmtpZ0XLN3dpb14GTr2XofIDI`
- I am using ngtcp2, and the reason is that it's by far the most buildable among QUIC libraries. It's a C library, so we have to be a little more careful with managing variables. But it's easy to build. You can build it in Ubuntu 26.04 and Debian without needing to manually compile any libraries. Since we already list OpenSSL as an optional dependency, I'm using that as the crypto backend (ngtcp2 supports multiple). You need `libngtcp2` and `libngtcp2_crypto_ossl`.
- You can set your server's private key and cert via `quic_cert_file` and `quic_key_file` in the config. By default it generates a self-signed certificate valid for 10 years.
- Everything is currently passed along a single bidirectional QUIC channel. I don't (yet) split the streams the way the original network implementation does.
- I did not add any advanced QUIC features yet. We can do this much later of course. So I did not do path migration (client wants to use a different IP address) or 0-RTT/retry tokens (honestly not really that useful for Luanti, more useful for HTTP).

**Comment by appgurueu** [comment]
This sounds pretty awesome. Would you mind opening a draft PR, or do you prefer another way of discussing the changes? Or should we just wait until a PR ready for review arrives?

**Comment by minlemon** [comment]
Good idea. I will open a draft PR once I've written some build instructions and cleaned it up a bit and then we can discuss. I'll need some feedback about quite a few things.

**Comment by minlemon** [comment]
After a lot more polishing and testing and explanations that I guess turned into a wall of text, here is the draft pull request: https://github.com/luanti-org/luanti/pull/17169

The CI isn't really accurate right now since I gated essentially the whole feature behind a CMake flag.
